### PR TITLE
tests: bootstrap manifests tests

### PR DESCRIPTION
**Pull request by fromanirh**
# Description

This patch adds the initial unit tests and their basic infrastructure.
Unit tests are mean to do simple, static checks, and to run (very)
often. Unit tests will check that configuration syntax is correct
(e.g. kernel args look sane, config file are parsable, base64-encoded
data can be decoded in something meaningful, any crossreference is
correct).

More elaborate testing (e.g. kernel tunables are actually applied,
isolated CPUs are really as such) will be done in e2e tests.

Each test should be put as close as possible to thing being tested,
using the `XXXX_test.go` naming convention.

Signed-off-by: Francesco Romani <fromani@redhat.com>

## Type of change

Please select the appropiate options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Self-testing
**Test Configuration**:

- Versions: N/A
- Hardware: N/A

## Checklist

- [X] I have performed a self-review of my own code
- [ ] If a change is adding a feature, it should require a change to the README.md and the review should catch this.
- [ ] If the change is a fix, it should have an issue. The review should make sure the comments state the issue (not just the number) and it should use the keywords that will close the issue on merge.
- [ ] A change should not being merged unless it passes CI or there is a comment/update saying what testing was passed.
- [ ] PRs should not be merged unless positively reviewed.

### CODE REVIEW

**Comment by fromanirh** [comment]
> Tests look great! thanks @fromanirh
> 
> I'd merge now but I don't want to conflict with @fedepaol's effort to create a top level `features` vendor folder with go.mod and go.sum that is consistent for the whole project.
> 
> The structure I was thinking we could use is similar to what OCS and KubeVirt did. They sprinkled unit tests all throughout the directories and created a single functional test package where all functional tests existed.
> 
> We'd have a single entry point `make unit-test` that would execute all the unit tests throughout the whole repo (minus the functional tests) and a `make func-test` that would execute just the tests in the functional test suite.

Thanks @davidvossel, you anticipated my next question. I like this approach, let's sync (/cc @fedepaol ) to have it asap!

File: features/performance/test/tuned_test.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+var _ = Describe("TestPerformanceTuned", func() {
+	table.DescribeTable("Tuned files should provide complete options",
+		func(fileName string) {
+			t := loadTuned(fileName)
+			Expect(t).ToNot(BeNil())
+			validateProfiles(fileName, t)
+		},
+		// msg, fileName
+		table.Entry(tunedYamls[0], tunedYamls[0]),
```
Comment (slintes) [inline features/performance/test/tuned_test.go]: doesn´t break anything, but the 2nd arg is unused?

Comment (fromanirh) [reply features/performance/test/tuned_test.go]: this is a sore point actually. The first argument is the title of the test table entry (e.g. "testing that FOO has BAR") the second is the test function argument. Let me try to clarify a bit.

Comment (slintes) [reply features/performance/test/tuned_test.go]: ah yes, sure, sorry. Should have known that. /me grabs another :coffee:

File: features/performance/test/machineconfig_test.go [hunk 2]
```diff
@@ -31,11 +36,29 @@ var _ = Describe("TestPerformanceMachineConfig", func() {
...
+			Expect(len(mc.Spec.Config.Storage.Files)).To(BeNumerically(">=", 1))
+			for _, encodedFile := range mc.Spec.Config.Storage.Files {
+				validateContentSource(encodedFile.Contents.Source)
+			}
+		})
+		It("Should ship at least an enabled systemd unit", func() {
+			mc := loadMC(machineconfigRTKernelYaml)
+			Expect(len(mc.Spec.Config.Systemd.Units)).To(BeNumerically(">=", 1))
+			for _, unitFile := range mc.Spec.Config.Systemd.Units {
+				Expect(*unitFile.Enabled).To(BeTrue())
```
Comment (MarSik) [inline features/performance/test/machineconfig_test.go]: I need to relearn golang, but the asterisk here is suspicious :) Why do the two lines differ wrt to it?

Comment (davidvossel) [reply features/performance/test/machineconfig_test.go]: Enabled is a boolean pointer.

to be slightly more accurate, we'd probably want this

```
Expect(unitFile.Enabled).ToNot(BeNil())
Expect(*unitFile.Enabled).To(BeTrue())
```

Comment (fromanirh) [reply features/performance/test/machineconfig_test.go]: @davidvossel yep, the more accurate the better. Changing.

**Comment by fromanirh** [comment]
depends on https://github.com/openshift-kni/baremetal-deploy/pull/56

**Comment by fromanirh** [comment]
please note this PR does *not* include hooking in the PR unittests flow intentionally. I'll file a separate PR for this, and the PR cover letter will include the rationale for this split.

**Comment by fromanirh** [comment]
/hold
not there yet. Tests currently assume the manifests are generated. They should instead generate them in their setup stage (and clean up the results)

**Comment by slintes** [comment]
> /hold
> not there yet. Tests currently assume the manifests are generated. They should instead generate them in their setup stage (and clean up the results)

isn't it enough to make the unitests make target depend on the generate target?

**Comment by fromanirh** [comment]
> > /hold
> > not there yet. Tests currently assume the manifests are generated. They should instead generate them in their setup stage (and clean up the results)
> 
> isn't it enough to make the unitests make target depend on the generate target?

It would be enough to test _something_, but that something isn't most likely what do we want to test.
Let me explain.

The tests I'm adding here are basic sanity check for manifests we generate. We can argue if they actually are (or can be easily made) unit tests, but let's scrape them for a moment. The key point here is that we want to check that the manifests produced by the generator don't contain obvious errors, e.g. can be accepted by the OCP components.
The actual behaviour will be covered by yet-to-be-written e2e tests, so we just care about lint-like check here.

But the generator has, and must have, parameters we can't predict. So the right approach is to do tabular testing and use the check that the manifests look good with the various parameters.

This is what will give us better benefits now and in the long term. It will however require few changes to the tests and to the generator scripts, I'm on it.

TL;DR: we can merge this PR, but code is not ready to be hooked under CI, will need some more work. (kudos to @fedepaol to point this up in another chat)

**Comment by fromanirh** [comment]
The PR should be done now. It will fail the `unittests` target and lane because
1. PR #56 not merged yet
2. deps not vendored (nor fetched on prow env) yet.

File: features/performance/go.mod [hunk 3]
```diff
@@ -0,0 +1,21 @@
+module github.com/openshift-kni/baremetal-deploy/performance
+
```
Comment (fedepaol) [inline features/performance/go.mod]: Would it make sense to have this consume the vendor folder at `/performance` level?
Also, not sure how go modules would behave having a nested module.

Comment (fromanirh) [reply features/performance/go.mod]: Dunno about nested modules, we'll see.
About consuming the `vendor` folder, that's up for discussion.

Comment (fedepaol) [reply features/performance/go.mod]: Correct me if I am wrong, but my understanding from what we discussed offline yestarday was that everything should leverage the `/feature` level go.mod / vendor folder.

Comment (fromanirh) [reply features/performance/go.mod]: Ah, sorry, I misunderstood your point. Then right, that's quoted aove a relic of a past version. I'll fix once your PR lands

Comment (fedepaol) [reply features/performance/go.mod]: It's already in: https://github.com/openshift-kni/baremetal-deploy/pull/55

File: features/performance/test/featuregate_test.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+// beware the typo (sic)
```
Comment (slintes) [inline features/performance/test/featuregate_test.go]: why not fixing the typo? :)

File: features/performance/test/kubeletconfig_test.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+			// parameters not really relevant for this test, just use something valid
+			kc := loadKubeletConfig(kubeletconfigYaml, "1-15", "0")
+			Expect(kc).ToNot(BeNil())
+			specKubeletConfig, err := decodeKubeletConfig(kc.Spec.KubeletConfig.Raw)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(specKubeletConfig.CPUManagerPolicy).To(Equal("static"))
+		})
+	})
+
+	// see: https://github.com/openshift-kni/baremetal-deploy/pull/56
```
Comment (slintes) [inline features/performance/test/kubeletconfig_test.go]: Can we remove the comment here? That PR is merged, and IMHO it's not important anymore to know about it

File: features/performance/test/machineconfig_test.go (hunk 2 above)
Comment (MarSik) [reply features/performance/test/machineconfig_test.go]: Ah, makes sense now. Thanks

File: features/performance/test/featuregate_test.go (hunk 4 above)
Comment (fromanirh) [reply features/performance/test/featuregate_test.go]: Yup, in a separate PR for better visibility and awareness

File: features/performance/test/kubeletconfig_test.go (hunk 5 above)
Comment (fromanirh) [reply features/performance/test/kubeletconfig_test.go]: Done

**Comment by fromanirh** [comment]
> /lgtm
> /approve

Thanks David!
